// fabric.js/src/coreapi/login/reducers.js
import {
    LOGIN_REQUEST,
    LOGIN_RESPONSE,
    LOGIN_EMAIL_INPUT,
    LOGIN_PASSWORD_INPUT,
    LOGIN_FORM_INVALID,
    LOGIN_FORM_VALID
} from './constants'

export function config(state = {}, action) {

    switch (action.type) {
        default:
            return state
    }
}

export function login(state = {
    isBusy: false,
    formIsValid: false,
}, action) {

    switch (action.type) {
        case LOGIN_FORM_VALID:
            return {
                ...state,
                formIsValid: true,
                formErrors: undefined,
            }
        case LOGIN_FORM_INVALID:
            return {
                ...state,
                formIsValid: false,
                formErrors: action.formErrors,
            }
        case LOGIN_REQUEST:
            return {
                ...state,
                isBusy: true,
                err: undefined,
                data: undefined,
            }
        case LOGIN_RESPONSE:
            return {
                ...state,
                isBusy: false,
                err: action.err,
                data: action.data,
            }
        case LOGIN_EMAIL_INPUT:
            return {
                ...state,
                email: action.email,
            }
        case LOGIN_PASSWORD_INPUT:
            return {
                ...state,
                password: action.password,
            }
        default:
            return state
    }
}
